addons/stock/models/stock_move.py

```diff
@@ -1014,10 +1014,12 @@ def _keys_out_sorted(ml):
     def _action_cancel(self):
         if any(move.state == 'done' for move in self):
             raise UserError(_('You cannot cancel a stock move that has been set to \'Done\'.'))
+        self = self.filtered(lambda x: x.state != 'cancel')
+        # self cannot contain moves that are either cancelled or done, therefore we can safely
+        # unlink all associated move_line_ids
+        self.mapped('move_line_ids').unlink()
```

Comment:
Would rather use self._do_unreserve() instead of unlink. _do_unreserve is multi and will also do only one unlink in batch. The rest of the PR is ok for me.

Reply:
OK, see nim's comment above for my "reasoning"